pkg/resource/deploy/step.go

```diff
@@ -953,23 +953,44 @@ func (s *ReadStep) Skip() {
 // resource by reading its current state from its provider plugin. These steps are not issued by the step generator;
 // instead, they are issued by the deployment executor as the optional first step in deployment execution.
 type RefreshStep struct {
-	deployment *Deployment                                // the deployment that produced this refresh
-	old        *resource.State                            // the old resource state, if one exists for this urn
-	new        *resource.State                            // the new resource state, to be used to query the provider
-	provider   plugin.Provider                            // the optional provider to use.
-	diff       plugin.DiffResult                          // the diff between the cloud provider and the state file
-	cts        *promise.CompletionSource[*resource.State] // the completion source to signal when the refresh is complete
+	// the deployment that produced this refresh
+	deployment *Deployment
+	// the old resource state, if one exists for this urn
+	old *resource.State
+	// the new resource state, to be used to query the provider
+	new *resource.State
+	// the URN to use for this refresh step, is taken from the new state
+	urn resource.URN
+	// the type of the resource to refresh, is taken from the new state
+	typ tokens.Type
+	// the provider to use for this refresh step, is taken from the new state
+	prov string
+	// the optional provider to use.
+	provider plugin.Provider
+	// the diff between the cloud provider and the state file
+	diff plugin.DiffResult
+	// the completion source to signal when the refresh is complete
+	cts *promise.CompletionSource[*resource.State]
 }
 
 // NewRefreshStep creates a new Refresh step.
-func NewRefreshStep(deployment *Deployment, cts *promise.CompletionSource[*resource.State], old *resource.State) Step {
+func NewRefreshStep(
+	deployment *Deployment, cts *promise.CompletionSource[*resource.State],
+	old *resource.State, new *resource.State,
+) Step {
 	contract.Requiref(old != nil, "old", "must not be nil")
+	if new == nil {
+		new = old
+	}
 
 	// NOTE: we set the new state to the old state by default so that we don't interpret step failures as deletes.
```

Comment:
Should this comment go above the `if` above?

Reply:
yeh makes sense